### PR TITLE
Simplify and improve the testing of the bag unpacker

### PR DESCRIPTION
**Pull request by alexwlchan**
* Pass around ObjectLocationPrefix in more places where it's appropriate (e.g. for “unpack this bag into this prefix”)
* Use an IngestID throughout the unpacker, rather than a String
* The Unpacker is now a type class with generic `get()` and `put()` methods (not a Store – they might point to different backends), and you create implementations of the Unpacker for the backends you want. This comes with our new generic test cases, so it’s more thoroughly tested.
* Archive has been completely rewritten to be simpler, to cut out the tail recursion and be easier to follow. It’s more generic (no UnpackSummary or T hanging around), and it has more test cases.

### CODE REVIEW

File: bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/services/Unpacker.scala [hunk 1]
```diff
@@ -62,77 +66,64 @@ case class Unpacker(s3Uploader: S3Uploader)(implicit s3Client: AmazonS3) {
 
     s"$archiveLocation could not be downloaded"
 
-//    cause.getStatusCode match {
-//      case 403 => s"access to $archiveLocation is denied"
-//      case 404 => s"$archiveLocation does not exist"
-//      case _   => s"$archiveLocation could not be downloaded"
-//    }
+    //    cause.getStatusCode match {
```
Comment (kenoir) [inline bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/services/Unpacker.scala]: delete?

File: bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/storage/Archive.scala [hunk 2]
```diff
@@ -2,118 +2,79 @@ package uk.ac.wellcome.platform.archive.bagunpacker.storage
...
+  * If you advance one, the other advances.  They move together.
+  *
+  *       +------------------------------------------+ original stream
+  *
+  *       +-------+                                    file1
+  *               +------------------+                 file2
+  *                                  +---------+       file3
+  *                                            +-----+ file4
+  *
+  * This is why we wrap the output in a close shield: if the caller closed
```
Comment (kenoir) [inline bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/storage/Archive.scala]: 🏆 Worthwhile Comment Award Winner 🏆

File: bag_unpacker/src/test/scala/uk/ac/wellcome/platform/archive/bagunpacker/storage/ArchiveTest.scala [hunk 3]
```diff
@@ -1,78 +1,101 @@
...
+import org.apache.commons.compress.archivers.ArchiveException
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry
+import org.apache.commons.compress.compressors.CompressorException
+import org.scalatest.{EitherValues, FunSpec, Matchers, TryValues}
+import uk.ac.wellcome.storage.streaming.Codec._
 
-import scala.util.Try
+import scala.util.Success
 
+/** The archive files used in these tests were deliberately created
```
Comment (kenoir) [inline bag_unpacker/src/test/scala/uk/ac/wellcome/platform/archive/bagunpacker/storage/ArchiveTest.scala]: 👏

File: bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/services/Unpacker.scala (hunk 1 above)
Comment (alexwlchan) [reply bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/services/Unpacker.scala]: Oh yeah, we commented this out as a reminder to go back and do it properly!

Done.

**Comment by alexwlchan** [comment]
Only failing test is a flaky bag replicator, so YOLO merging.

**Pull request merged**